tests/unit/request/xhr.ts

```diff
@@ -2,168 +2,497 @@ import * as registerSuite from 'intern!object';
+	}
+}
+
+let echoServerAvailable = false;
 registerSuite({
 	name: 'request/xhr',
 
+	before: function () {
+		return new Promise(function (resolve, reject) {
+			xhrRequest('/__echo/', {
+				method: 'get',
+				timeout: 10000
+			}).then(
+				function (response) {
+					if (response && response.statusCode === 200) {
+						echoServerAvailable = true;
+					}
+					resolve();
+				},
+				function () {
+					resolve();
+				}
+			);
+		});
+	},
+
 	'HTTP methods': {
-		specified(): void {
-			const dfd = this.async();
-			xhrRequest(getRequestUrl('foo.json'), { method: 'get' })
+		specified() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest(getRequestUrl('foo.json'), { method: 'get' })
 				.then(
-					dfd.callback(function (response: any): void {
+					function (response: any) {
 						assert.strictEqual(response.requestOptions.method, 'get');
-					}),
-					dfd.reject.bind(dfd)
+					}
 				);
 		},
 
-		'default'(): void {
-			const dfd = this.async();
-			xhrRequest(getRequestUrl('foo.json'))
+		'default'() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest(getRequestUrl('foo.json'))
 				.then(
-					dfd.callback(function (response: any): void {
+					function (response: any) {
 						assert.strictEqual(response.requestOptions.method, 'GET');
-					}),
-					dfd.reject.bind(dfd)
+					}
 				);
+		},
+
+		'.get with URL query'() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest('/__echo/xhr?color=blue&numbers=one&numbers=two', {
+				handleAs: 'json'
+			}).then(
+				function (response: any) {
+					const query = JSON.parse(response.data).query;
+					assert.strictEqual(query.color, 'blue');
+					assert.deepEqual(query.numbers, [ 'one', 'two' ]);
+					assert.strictEqual(
+						response.url,
+						'/__echo/xhr?color=blue&numbers=one&numbers=two'
+					);
+				}
+			);
+		},
+
+		'.post': function () {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest('/__echo/post', {
+				method: 'POST',
+				data: new UrlSearchParams({ color: 'blue' }).toString(),
+				responseType: 'json'
+			}).then(
+				function (response: any) {
+					const data = JSON.parse(response.data);
+					assert.strictEqual(data.method, 'POST');
+					const payload = data.payload;
+
+					assert.ok(payload && payload.color);
+					assert.strictEqual(payload.color, 'blue');
+				}
+			);
 		}
 	},
 
 	'request options': {
-		'timeout'(): void {
-			const dfd = this.async();
-			xhrRequest(getRequestUrl('foo.json'), { timeout: 1 })
+		'timeout'() {
+			if (!echoServerAvailable) {
+				this.skip('No echo server available');
+			}
+			return xhrRequest(getRequestUrl('foo.json'), { timeout: 1 })
 				.then(
-					dfd.resolve.bind(dfd),
-					dfd.callback(function (error: Error): void {
+					null,
```

Comment:
Should it throw a RequestTimeoutError? If it throws something else the test fails because the error.name doesn't match in the error handler. But if we're throwing that explicitly what would this really be testing?